### PR TITLE
add finalizer for tsp controller

### LINKED ISSUE #188: Add finalizers for TimeSeriesPrediction controller
## Describe the feature

Avoid the missed event for time series prediction, add finalizers for time series prediction

### PR DESCRIPTION
**Pull request by kitianFresh**
#### What type of PR is this?
Feature 

#### What this PR does / why we need it:
Add finalizers for time series prediction to avoid missing delete event, clear prediction cache

#### Which issue(s) this PR fixes:

Fixes #188

### CODE REVIEW

File: pkg/controller/timeseriesprediction/status.go [hunk 1]
```diff
@@ -150,7 +150,7 @@ func (tc *Controller) doPredict(tsPrediction *predictionapi.TimeSeriesPrediction
 		if klog.V(6).Enabled() {
 			apiDataBytes, err1 := json.Marshal(predictedData)
 			dataBytes, err2 := json.Marshal(data)
-			klog.V(6).Infof("DoPredict predicted data details, key: %v, queryExpr: %v, apiData: %v, predictData: %v, errs: %+v", klog.KObj(tsPrediction), queryExpr, string(apiDataBytes), string(dataBytes), []error{err1, err2})
+			klog.V(6).Infof("DoPredict predicted data detaiensurePredictionFinalizer.ls, key: %v, queryExpr: %v, apiData: %v, predictData: %v, errs: %+v", klog.KObj(tsPrediction), queryExpr, string(apiDataBytes), string(dataBytes), []error{err1, err2})
```
Comment (qmhu) [inline pkg/controller/timeseriesprediction/status.go:153]: typo?

File: pkg/controller/timeseriesprediction/time_series_prediction_controller.go [hunk 2]
```diff
@@ -54,18 +59,32 @@ func (tc *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Res
...
-			klog.V(4).Infof("Failed to load exist tsp %v", req.String())
-			return ctrl.Result{}, nil
+			klog.V(4).Infof("Failed to load cached tsp %v, maybe the craned is reboot", req.String())
+			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
 		klog.V(4).Infof("Failed to get TimeSeriesPrediction %v err: %v", klog.KObj(p), err)
 		return ctrl.Result{Requeue: true}, err
 	}
 
+	if !p.ObjectMeta.GetDeletionTimestamp().IsZero() {
```
Comment (qmhu) [inline pkg/controller/timeseriesprediction/time_series_prediction_controller.go:74]: `if p.DeletionTimestamp != nil`

Comment (kitianFresh) [reply pkg/controller/timeseriesprediction/time_series_prediction_controller.go:74]: > `if p.DeletionTimestamp != nil`

it is ok when the time is nil, the IsZero will judge nil pointer and return true

File: pkg/controller/timeseriesprediction/status.go (hunk 1 above)
Comment (kitianFresh) [reply pkg/controller/timeseriesprediction/status.go:153]: > typo?

oh yeah. fix it

**Comment by mfanjie** [comment]
per discussion offline, we can add finalizer later if actual memory leak is found.

**Pull request closed without merging**